### PR TITLE
Update the API for forward() and add corresponding tutorials

### PR DESCRIPTION
**Pull request by TianyuDu**

### CODE REVIEW

**Review by TianyuDu — COMMENTED** [review]
reviewed by Tianyu.

File: bemb/model/bemb.py [hunk 1]
```diff
@@ -5,13 +5,13 @@
...
 from pprint import pprint
-from typing import Dict, List, Optional, Union, Tuple
+from typing import Dict, List, Optional, Tuple, Union
 
 import numpy as np
 import torch
 import torch.nn as nn
 from torch_choice.data import ChoiceDataset
-from torch_scatter import scatter_max, scatter_logsumexp
+from torch_scatter import scatter_logsumexp, scatter_max
```
Comment (kanodiaayush) [inline bemb/model/bemb.py:14]: have you added automatic linting/styling? this seems to be a change for alphabetical order, for eg.

do I/ a new user has to make sure of this before committing new code?

File: bemb/model/bemb.py [hunk 2]
```diff
@@ -405,6 +413,41 @@ def sample_log_likelihoods(self, batch:ChoiceDataset, sample_dict: Dict[str, tor
...
+    def predict_proba(self, batch: ChoiceDataset) -> torch.Tensor:
+        """
+        Draw prediction on a given batch of dataset.
+
+        Args:
+        batch (ChoiceDataset): the dataset to draw inference on.
+
+        Returns:
+        torch.Tensor: the predicted probabilities for each class, the behavior varies by self.pred_item.
+        (1: pred_item == True) While predicting items, the return tensor has shape (len(batch), num_items), out[i, j] is the predicted probability for choosing item j AMONG ALL ITEMS IN ITS CATEGORY in observation i. Please note that since probabilities are computed from within-category normalization, hence out.sum(dim=0) can be greater than 1 if there are multiple categories.
```
Comment (kanodiaayush) [inline bemb/model/bemb.py:426]: may not be a big deal, but I have not seen this out.* format

File: bemb/model/bemb.py [hunk 3]
```diff
@@ -405,6 +413,41 @@ def sample_log_likelihoods(self, batch:ChoiceDataset, sample_dict: Dict[str, tor
...
+        (2: pred_item == False) While predicting external labels for each observations, out[i, 0] is the predicted probability for label == 0 on the i-th observation, out[i, 1] is the predicted probability for label == 1 on the i-th observation. Generally, out[i, 0] + out[i, 1] = 1.0. However, this could be false if under-flowing/over-flowing issue is encountered.
+        We highly recommend users to use the forward function to get the log-prob instead.
```
Comment (kanodiaayush) [inline bemb/model/bemb.py:428]: you should rephrase this as
We highly recommend users to get log-probs as those are less prone to overflow/underflow; those can be accessed using the forward() function

File: notes_on_refreshing_api.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Towards BEMB `v1.0`.
+We are planning to refine and expand the current API of `BEMBFlex`.
+## The `pred_item` and multiple class prediction.
+- [ ] Currently the model supports predicting *binary* `batch.label` or multi-class `batch.item_index`. We plan to support arbitrary multi-class classifications.
+    * In particular, you don't need to change anything if `pred_item=True`, the model will know the number of classes is exactly the `num_items` parameter. Also, in this case, your `ChoiceDataset` object does **not** need to have a `label` attribute, since the model will look for the `item_index` as the ground truth for training.
+    * In contrast, if `pred_item=False`, now you need to supply a `num_classes` to the `BEMBFlex.__init__()` method. Also, you would need a `label` attribute in the `ChoiceDataset` object. The `label` attribute should be a `LongTensor` with values from `{0, 1, ..., num_classes}`.
```
Comment (kanodiaayush) [inline notes_on_refreshing_api.md:6]: in more supermarkets, we have utils 0.0, 0.3, 0.5 and 1.0; should we allow float values here? I think we should. But that makes it complicated because that is then treating these as values with ordering rather than independent numbers.

File: notes_on_refreshing_api.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+## Post-Estimation
+- [ ] Thanks to feedbacks from our valued users, we are planning to reorganize our post-estimation prediction methods for better user experience.
+    - [ ] We will implement a method called `predict_proba()`, the same name as inference methods of scikit-learn models.
+    - [ ] This method will have `@torch.no_grad()` as a decorator, so you can use it however you want without being worried about gradient tracking.
+    - [ ] With `pred_items = True`, the `batch` needs `item_index` attribute only if it's involved in the utility computation (e.g., within-category computation).
```
Comment (kanodiaayush) [inline notes_on_refreshing_api.md:12]: i didn't understand this point; also it might not be worth it but `pred_item` feels like an uninformative label; maybe `pred_between_items` is better (can't even use multinomial choice because we are not planning to support multinomial choice for `pred_item = False`

Comment (kanodiaayush) [reply notes_on_refreshing_api.md:12]: I'm just noting this; i am pretty sure you don't want to relabel this everywhere

File: tests/bemb_functionality.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+"""
```
Comment (kanodiaayush) [inline tests/bemb_functionality.py:1]: I'm not checking these in detail; just taking an overview

File: tests/compare_bemb_models.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
+"""
+Compares the parameter estimation from BEMB in torch and C++.
```
Comment (kanodiaayush) [inline tests/compare_bemb_models.py:2]: this is a good test. thanks for adding this.

File: tutorials/hyperparameter_tuning_with_ray/tune_model.ipynb [hunk 8]
```diff
@@ -631,12 +631,12 @@
...
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.10"
+   "version": "3.9.7"
   },
   "orig_nbformat": 4,
   "vscode": {
    "interpreter": {
-    "hash": "07dd84f3285103b45d36665747e055dbb0ef78c9d17293fd1b33b99205a40b05"
+    "hash": "5859d33511df864b0b7226a715510a0165ef032ed4b83eb4ae2c092f0788759c"
```
Comment (kanodiaayush) [inline tutorials/hyperparameter_tuning_with_ray/tune_model.ipynb:639]: vat is this thing???

**Review by kanodiaayush — COMMENTED** [review]
general points

i have not reviewed the ipynb files; I trust that that is okay
have you added tests for movie lens already? I don't think so, right?

also; I feel like I hope I never have to update to new bemb again and I can keep the one I have for the more supermarkets project. else, that is going to be a HUGE update!!

File: bemb/model/bemb.py (hunk 1 above)
Comment (TianyuDu) [reply bemb/model/bemb.py:14]: I was using the automatic "sort import statements" function in the VSCode, it seems to be the alphabetical order after the commit since "L" goes before "M".

File: bemb/model/bemb.py (hunk 2 above)
Comment (TianyuDu) [reply bemb/model/bemb.py:426]: Yeah, this is just to provide users some ideas on what is returned.

File: bemb/model/bemb.py (hunk 3 above)
Comment (TianyuDu) [reply bemb/model/bemb.py:428]: rephased, I pushed to the main branch directly.

File: tutorials/hyperparameter_tuning_with_ray/tune_model.ipynb (hunk 8 above)
Comment (TianyuDu) [reply tutorials/hyperparameter_tuning_with_ray/tune_model.ipynb:639]: Hmmm, I don't really know. It looks like something that the Jupyter notebook is maintaining internally.
